ENH add gap safe screening rules to enet_coordinate_descent_gram

lorentzenchr opened the pull request:
#### Reference Issues/PRs
Partially solves solves #229.
Continuation of #31882 (which should be merged first!!!)

#### What does this implement/fix? Explain your changes.
Gap safe screening rules for the Cython funtion `sparse_enet_coordinate_descent`.

#### Any other comments?
A bit of refactoring for redundant code.

Also positively effects
- covariance: `GraphicalLasso(mode="cd")`, `GraphicalLassoCV(mode="cd")`, `graphical_lasso_path(mode="cd")`
- decomposition: `SparsePCA(method="cd")`, `MiniBatchSparsePCA(method="cd")`, `DictionaryLearning(fit_algorithm="cd")`, `MiniBatchDictionaryLearning(fit_algorithm="cd")`, `dict_learning_online`, `dict_learning`, `sparse_encode`

lorentzenchr committed:
MNT rename ii -> j

lorentzenchr committed:
ENH add gram_gap_enet and check stopping before main loop

lorentzenchr committed:
ENH add gap safe screening rule for enet_coordinate_descent_gram

lorentzenchr commented:
> There are some tests (dict learning and graphical lasso) 

I have seen it. Requires a deeper digging, maybe the new early check of the stopping criterion...

lorentzenchr committed:
TST improve/fix test_cython_solver_equivalence

lorentzenchr committed:
TST fix sparse pca and dictionary lerning in sklearn.decomposition

lorentzenchr committed:
FIX early stopping in CD only when 0 <= gap

lorentzenchr committed:
TST adapt/fix test_graphical_lasso.py

lorentzenchr committed:
ENH improve max_iter of inner CD solver in _graphical_lasso

lorentzenchr committed:
CLN better if condition for Qw update

lorentzenchr committed:
TST fix rtol of test_dict_learning_numerical_consistency

lorentzenchr committed:
MNT loosen rtol in test_fit_transform

lorentzenchr committed:
CLN rename gram_gap_enet to gap_enet_gram

lorentzenchr committed:
FIX enet_coordinate_descent_gram

lorentzenchr committed:
TST loosen atol in test_graphical_lassos

lorentzenchr committed:
Merge branch 'main' into gap_safe_3_gram

lorentzenchr committed:
TST 2nd take on fixing test_fit_transform_tall and test_graphical_lassos

lorentzenchr committed:
Merge branch 'main' into gap_safe_3_gram

lorentzenchr committed:
FIX graphical_lasso mode="cd" now uses cyclic instead of random CD

* The random CD caused random behaviour because rng=check_random_state(None)
* Locally all tests pass with
SKLEARN_TESTS_GLOBAL_RANDOM_SEED="all" pytest -x -Werror sklearn/covariance/tests/test_graphical_lasso.py::test_graphical_lassos

lorentzenchr committed:
DOC add whatsnew for the fixed GraphicalLasso

lorentzenchr committed:
DOC extend whatsnew entry 31882 for gap safe screening rules

lorentzenchr committed:
CLN remove comment about unstable test with different random seeds

lorentzenchr committed:
run tests with [all random seeds]

lorentzenchr committed:
DOC add whatsnew for efficiency improvement of graphical lasso

lorentzenchr committed:
DOC add another whatsnew entry for decomposition module

virchan commented on sklearn/covariance/_graph_lasso.py:
```diff
@@ -138,16 +138,23 @@ def _graphical_lasso(
+                            Q=sub_covariance,
+                            q=row,
+                            y=row,
+                            # TODO: It is not ideal that the max_iter of the outer
+                            # solver (graphical lasso) is coupled with the max_iter of
+                            # the inner solver (CD). Ideally, CD has its own parameter
+                            # enet_max_iter (like enet_tol). A minimum of 20 is rather
+                            # arbitrary, but not unreasonable.
```
Naive question: Do you think we should mention this in the [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.covariance.GraphicalLasso.html#sklearn.covariance.GraphicalLasso) as well? I feel like the end-users should be aware of this.

lorentzenchr replied:
For me, that’s an implementieren detail. I would rather open an issue for it because we should either expose a parameter for it or find a smart automatic setting.

lorentzenchr replied:
I opened #32109.

lorentzenchr committed:
Merge branch 'main' into gap_safe_3_gram

virchan commented on doc/whats_new/upcoming_changes/sklearn.linear_model/31986.efficiency.rst:
```diff
@@ -1,12 +1,10 @@
   loop of descent steps. As the stopping criterion requires the computation of the dual
   gap, the screening happens whenever the dual gap is computed.
-  By :user:`Christian Lorentzen <lorentzenchr>` :pr:`31882` and
+  By :user:`Christian Lorentzen <lorentzenchr>` :pr:`31882`, :pr:`31987` and
```
Yea, false alarm on my side. Thanks catching that.

lorentzenchr commented:
Dear reviewers, feel free to finish in case I am unresponsive.

virchan reviewed (approved):
LGTM! Thanks, @lorentzenchr!

@ogrisel, would you like to have a look and merge this?

ogrisel committed:
Trigger CI for all tests that might be impacted with [all random seeds]
test_graphical_lassos
test_graphical_lasso_when_alpha_equals_0
test_graphical_lasso_cv
test_graphical_lasso_cv_scores
test_graphical_lasso_cv_scores_with_routing
test_fit_transform
test_transform_nan
test_fit_transform_tall
test_initialization
test_scaling_fit_transform
test_pca_vs_spca
test_spca_early_stopping
test_equivalence_components_pca_spca
test_sparse_pca_inverse_transform
test_transform_inverse_transform_round_trip

ogrisel reviewed (commented):
I pushed an empty commit (7263e37) with a detailed `[all random seeds]` message to ensure that this PR does not introduce seed sensitivity in potentially impacted tests.

Assuming they are green, and the changelog conflict is resolved, LGTM on my side. Just two minor comments below. Feel free to merge if I am not responsive.

ogrisel reviewed (approved):
I meant to "approve" rather than "comment" in the previous review.

ogrisel committed:
Remove useless line.

ogrisel committed:
Merge branch 'main' into gap_safe_3_gram
